[Draft] Move specification from ROS wiki to the urdfdom repo (#183)
ros/urdfdom: specification/sensor.md

# `<sensor>` element

>[!WARNING]

> > The sensor element has been implemented in the URDF Dom but has never really been used in application. This is a project that was dropped anyone is encouraged to pick it up and extend it to sensor hardware applications. Please contribute!

Reviewer: ```suggestion
> The sensor element has been implemented in the URDF DOM but has never really been used in application. This is a project that was dropped. Anyone is encouraged to pick it up and extend it to sensor hardware applications. Please contribute!

```

Reviewer 2: I don't think this warning is actually something we should have in the specification.  That is, we have support for sensors, so we should just say that.  The fact that no downstreams that we know of use it is a deployment/implementation detail, not a specification detail.

> >[!WARNING]
> > The sensor element has been implemented in the URDF Dom but has never really been used in application. This is a project that was dropped anyone is encouraged to pick it up and extend it to sensor hardware applications. Please contribute!

Author: ```suggestion
```

The sensor element describes basic properties of a visual sensor (i.e. camera / ray sensor).

Here is an example of a camera sensor element:
```xml
<sensor name="my_camera_sensor" update_rate="20">
    <parent link="optical_frame_link_name"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <camera>
        <image width="640" height="480" hfov="1.5708" format="RGB8" near="0.01" far="50.0"/>
    </camera>
</sensor>
```

And below is an example of a laser scan (ray) sensor element:

```xml
<sensor name="my_ray_sensor" update_rate="20">
    <parent link="optical_frame_link_name"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <ray>
        <horizontal samples="100" resolution="1" min_angle="-1.5708" max_angle="1.5708"/>
        <vertical samples="1" resolution="1" min_angle="0" max_angle="0"/>
    </ray>
</sensor>

> ```

Reviewer 2: In most of the other files, the Examples are *after* the rest of the specification.  I think that makes sense; talk about the individual items, and then show how to put it together.  I'll suggest the same here.

Author: Fixed in https://github.com/ros/urdfdom/pull/183/commits/0ae2dca39977796bdba6227ec66158fad7d4c466 .

## Attributes

| attribute     | type     | use      | default value                                                                                                                | description                 |
| ------------- | -------- | -------- | ---------------------------------------------------------------------------------------------------------------------------- | --------------------------- |
| `name`        | `string` | required | NA                                                                                                                           | The name of the sensor link |
| `update_rate` | `double` | optional | The frequency at which the sensor data is generated in `hz`. If left unspecified, the sensor will generate data every cycle. |

## Elements

| element                | use      | description                                                                                  |
| ---------------------- | -------- | -------------------------------------------------------------------------------------------- |

> | [`<parent`](#parent)   | required | Defines the parent link this sensor is attached to.                                          |

Reviewer: ```suggestion
| [`<parent>`](#parent)   | required | Defines the parent link this sensor is attached to.                                          |

```

> | [`<origin>`](#origin)` | optional | This is the pose of the sensor optical frame, relative to the sensor parent reference frame. |

Reviewer: ```suggestion
| [`<origin>`](#origin) | optional | This is the pose of the sensor optical frame, relative to the sensor parent reference frame. |

```

| [`<camera>`](#camaera) | optional | Camera sensor definition.                                                                    |
| [`<ray>`](#ray)        | optional | Laser sensor definition.                                                                     |

### `<parent>`

| attribute | type     | use      | default value | description                                             |
| --------- | -------- | -------- | ------------- | ------------------------------------------------------- |
| `link`    | `string` | required | NA            | The name of the parent link this sensor is attached to. |

### `<origin>`

> | attribute     | type     | use      | default value                                                                                                                | description                 |

Reviewer: The table doesn't render correctly.

Author: Fixed in https://github.com/ros/urdfdom/pull/183/commits/0ae2dca39977796bdba6227ec66158fad7d4c466 .

| `xyz` | `string` | optional | zero vector | Represents the offset with respect to the parent frame. |
| `rpy` | `string` | optional | zero vector | Represents the fixed axis roll, pitch and yaw angles in radians. |

> ### `<camaera>`

Reviewer: ```suggestion
### `<camera>`

```

| -------------------------- | -------- | ----------------------------- |
| [`<image>`](#camera-image) | required | Defines the image parameters. |

#### camera: `<image>`

| attribute | type           | use      | default value | description                                                                                                                                                                                        |
| --------- | -------------- | -------- | ------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `width`   | `unsigned int` | required | NA            | Width of the camera in `pixels`.                                                                                                                                                                   |
| `height`  | `unsigned int` | required | NA            | Height of the camera in `pixels`.                                                                                                                                                                  |

> | `format`  | `string`       | required | NA            | Can be any of the strings defined in [image_encodings.h insdie sensor_msgs](https://code.ros.org/trac/ros-pkg/browser/stacks/common_msgs/trunk/sensor_msgs/include/sensor_msgs/image_encodings.h). |

Reviewer: ```suggestion
| `format`  | `string`       | required | NA            | Can be any of the strings defined in [image_encodings.h inside sensor_msgs](https://github.com/ros/common_msgs/blob/noetic-devel/sensor_msgs/include/sensor_msgs/image_encodings.h). |

```

| `samples`    | `unsigned int` | optional | 1                                                                   | The number of simulated rays to generate per complete laser sweep cycle.                                                                                                                                           |
| `resolution` | `unsigned int` | optional | 1                                                                   | This number is multiplied by samples to determine the number of range data points returned. If resolution is less than one, range data is interpolated. If resolution is greater than one, range data is averaged. |
| `min_angle`  | `double`       | optional | Minimun angle in `radians`.                                         |
| `max_angle`  | `double`       | optional | Maximum angle in `radian`. Must be greater or equal to `min_angle`. |

## Recommended Camera or Ray Resolution

In simulation, large sensors will slow down overall performance. Depending on update rates required, it is recommended to keep the camera or ray resolution and update rates as low as possible.

## Proposal for New Type of Sensor

> TBD

Reviewer 2: Just remove this, it only belongs in the specification once it has been implemented.

Author: ```suggestion
```
